File: surf-api-bukkit/surf-api-bukkit-server/src/main/java/dev/slne/surf/surfapi/bukkit/server/scoreboard/SurfAutoUpdatableScoreboardImpl.java
package dev.slne.surf.surfapi.bukkit.server.scoreboard;

import dev.slne.surf.surfapi.bukkit.api.scoreboard.SurfAutoUpdatableScoreboard;
import dev.slne.surf.surfapi.bukkit.server.BukkitMain;
import net.kyori.adventure.text.Component;
import net.megavex.scoreboardlibrary.api.sidebar.component.SidebarComponent;
import net.megavex.scoreboardlibrary.api.sidebar.component.animation.SidebarAnimation;
import org.bukkit.scheduler.BukkitRunnable;
import org.jetbrains.annotations.ApiStatus;

import java.util.List;

@ApiStatus.Internal
public class SurfAutoUpdatableScoreboardImpl extends SurfScoreboardImpl implements SurfAutoUpdatableScoreboard {

    private BukkitRunnable updater;

    public SurfAutoUpdatableScoreboardImpl(Component title, int maxLines, SidebarComponent sidebarComponent, List<SidebarAnimation<Component>> animations) {
        super(title, maxLines, sidebarComponent, animations);
    }

    @Override
    public void enable() {
        super.enable();
        setupUpdater();
        updater.runTaskTimerAsynchronously(BukkitMain.getInstance(), 0, 5);
    }

    @Override
    public void disable() {
        updater.cancel();
        super.disable();
    }

    private void setupUpdater() {
        updater = new BukkitRunnable() {
            @Override
            public void run() {
                update();
            }
        };
    }
}

File: surf-api-bukkit/surf-api-bukkit-server/src/main/kotlin/dev/slne/surf/surfapi/bukkit/server/impl/scoreboard/SurfAutoUpdatableScoreboardImpl.kt
package dev.slne.surf.surfapi.bukkit.server.impl.scoreboard

import com.github.shynixn.mccoroutine.folia.launch
import com.github.shynixn.mccoroutine.folia.ticks
import dev.slne.surf.surfapi.bukkit.api.scoreboard.ObsoleteScoreboardApi
import dev.slne.surf.surfapi.bukkit.api.scoreboard.SurfAutoUpdatableScoreboard
import dev.slne.surf.surfapi.bukkit.server.plugin
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import net.kyori.adventure.text.Component
import net.megavex.scoreboardlibrary.api.sidebar.component.SidebarComponent
import net.megavex.scoreboardlibrary.api.sidebar.component.animation.SidebarAnimation

@ObsoleteScoreboardApi
open class SurfAutoUpdatableScoreboardImpl(
    title: Component,
    maxLines: Int,
    sidebarComponent: SidebarComponent,
    animations: List<SidebarAnimation<Component>>
) : SurfScoreboardImpl(title, maxLines, sidebarComponent, animations), SurfAutoUpdatableScoreboard {
    private var updater: Job? = null

    override fun enable() {
        super.enable()

        this.updater = launchUpdater()
    }

    override fun disable() {
        super.disable()

        updater!!.cancel()
    }

    private fun launchUpdater() = plugin.launch {
        while (true) {
            update()
            delay(5.ticks)
        }
    }
}
